providers/axfrddns/axfrddnsProvider.go

```diff
@@ -0,0 +1,328 @@
+	}
+	// TODO ensure that secret is valid base64
+	return &Key{algo: algo, id: arr[1] + ".", secret: arr[2]}, nil
+}
+
+type Param struct {
+	DefaultNS []string `json:"default_ns"`
+}
+
+type Key struct {
+	algo   string
+	id     string
+	secret string
+}
+
+type AxfrDdns struct {
+	rand        *rand.Rand
+	master      string
+	nameservers []*models.Nameserver
+	transferKey *Key
+	updateKey   *Key
+}
+
+// GetNameservers returns the nameservers for a domain.
+func (c *AxfrDdns) GetNameservers(domain string) ([]*models.Nameserver, error) {
+	return c.nameservers, nil
+}
+
+// GetZoneRecords gets the records of a zone and returns them in RecordConfig format.
+func (c *AxfrDdns) GetZoneRecords(domain string) (models.Records, error) {
+
+	transfer := new(dns.Transfer)
+	transfer.DialTimeout = dnsTimeout
+	transfer.ReadTimeout = dnsTimeout
+	request := new(dns.Msg)
+	request.SetAxfr(domain + ".")
+
+	if c.transferKey != nil {
+		transfer.TsigSecret =
+			map[string]string{c.transferKey.id: c.transferKey.secret}
+		request.SetTsig(c.transferKey.id, c.transferKey.algo, 300, time.Now().Unix())
+	}
+
+	envelope, err := transfer.In(request, c.master)
+	if err != nil {
+		return nil, err
+	}
+
+	var rawRecords []dns.RR
+	for msg := range envelope {
+		if msg.Error != nil {
+			// Fragile but more "user-friendly" error-handling
+			if msg.Error.Error() == "dns: bad xfr rcode: 9" {
+				return nil, fmt.Errorf("dns: NOT AUTH")
+			}
+			return nil, msg.Error
+		}
+		rawRecords = append(rawRecords, msg.RR...)
+	}
+
+	var foundDNSSecRecords *models.RecordConfig
+	foundRecords := models.Records{}
+	for _, rr := range rawRecords {
+		switch rr.(type) {
+		case *dns.RRSIG,
+			*dns.DNSKEY,
+			*dns.CDNSKEY,
+			*dns.CDS,
+			*dns.NSEC,
+			*dns.NSEC3,
+			*dns.NSEC3PARAM:
+			// Ignoring DNSSec RRs, but replacing it with a single
+			// "TXT" placeholder
+			if foundDNSSecRecords == nil {
+				foundDNSSecRecords = new(models.RecordConfig)
+				foundDNSSecRecords.Type = "TXT"
+				foundDNSSecRecords.SetLabel(dnssec_dummy_label, domain)
+				err = foundDNSSecRecords.SetTargetTXT(dnssec_dummy_txt)
+				if err != nil {
+					return nil, err
+				}
+			}
+			continue
+		default:
+			rec := models.RRtoRC(rr, domain)
+			foundRecords = append(foundRecords, &rec)
+		}
+	}
+
+	if len(foundRecords) >= 1 && foundRecords[len(foundRecords)-1].Type == "SOA" {
+		// When TSig is used, the SOA is sent two times: as the
+		// first and the last record
+		foundRecords = foundRecords[:len(foundRecords)-1]
+	}
+
+	if foundDNSSecRecords != nil {
+		foundRecords = append(foundRecords, foundDNSSecRecords)
+	}
+
+	return foundRecords, nil
+}
+
+// GetDomainCorrections returns a list of corrections to update a domain.
+func (c *AxfrDdns) GetDomainCorrections(dc *models.DomainConfig) ([]*models.Correction, error) {
+	dc.Punycode()
+
+	foundRecords, err := c.GetZoneRecords(dc.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(foundRecords) >= 1 && foundRecords[0].Type == "SOA" {
```

Comment:
Are we sure that the SOA will always be the first record?  (I presume so, otherwise the "S" wouldn't meant "start" but... thought I'd check)

Reply:
Yes, this is a "MUST". See section 2.2 of rfc5936.

But, if you prefer, I can add some protective programming. In case of a buggy servers don't follow the RFC.

Reply:
I added a comment in the code.